.github/workflows/deployment.yml

```diff
@@ -0,0 +1,158 @@
+name: Deployment
+
+# The deployment workflow should only run when changes are merged into the `main` branch.
+# Since a branch protection rule prevents directly pushing to `main` this workflow will
+# only run on a successful merge request.
+on:
+  push:
+    branches:
```

Comment:
Why not trigger on releases as well?

Reply:
I'm not sure I understand how that would work. Can you provide some more details? The `deployment` workflow is responsible for creating a release. Would that not trigger the workflow again?

Reply:
Oh I see, I missed that at the bottom. 

I generally create the release (+ tag) from the GitHub GUI as it auto-populates information such as commit diffs etc and generally we would have to go back and edit the release notes anyways. Additionally, GitHub provides hooks to limit the workflow specifically to when the release is published.

Here's an example: https://github.com/opendcs/rest_api/blob/main/.github/workflows/publish.yml#L4

Reply:
I think our deployment scenario is a little different than the one you linked to above. There it looks like your only deployment target is a GitHub Release. In our case we actually have two different deployment flows:

1. Any time a pull request is merged into `main` we publish to TestPyPI. This way we have a development/snapshot package that can be used for testing.
2. When a tagged pull request is merged into `main` we publish to both TestPyPI and PyPI (updating both the testing and release packages), and we also create a GitHub Release.

The main difference is that we want to do the first case whenever changes are merged into `main`, not just when a release is published.

It should be possible to manually trigger the `deployment` workflow, but I'm not sure how that compares to creating a release manually using the GitHub UI.

Reply:
yeah, we've generally stopped publishing snapshots on main in a bunch of HEC projects because in practice we never end up needing them (an exceptions, we trigger manual deployments). Different here given that a python library targets a larger scripting audience compared to enterprise software deployments so that makes sense.